src/App.tsx

```diff
@@ -1,24 +1,239 @@
+import { TodoItem } from './Components/TodoItem';
+import { Footer } from './Components/Footer';
+import { Header } from './Components/Header';
+import { ErrorNotification } from './Components/ErrorNotification';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [condition, setCondition] = useState(ForComletedTodo.All);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const [inputDisabled, setInputDisabled] = useState(false);
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorTimer, setErrorTimer] = useState<NodeJS.Timeout | null>(null);
+
+  useEffect(() => {
+    if (errorTimer) {
+      clearTimeout(errorTimer);
+    }
+
+    setErrorTimer(setTimeout(() => {
+      setErrorMessage(null);
+    }, 3000));
+  }, [errorMessage, errorTimer]);
+
+  const hasTodos = !!todos.length;
+
+  const isAllCompleted = todos.every(todo => todo.completed);
+
+  const fetchData = useCallback(async () => {
+    try {
+      setErrorMessage(null);
+      const todosFetch = await todoService.getTodos();
+
+      setTodos(todosFetch);
+    } catch (err) {
+      setErrorMessage('Unable to load todos');
+    }
+  }, []);
+
+  useEffect(() => {
+    fetchData();
+  }, []);
+
+  const hadleAddTodo = async (title: string) => {
+    try {
+      setInputDisabled(true);
+
+      const newTodo = {
+        id: 0,
+        userId: 0,
+        title: title.trim(),
+        completed: false,
+      };
+
+      setTempTodo(newTodo);
+
+      const addedTodo = await todoService.addTodo(title);
+
+      setTodos((prevTodos) => [...prevTodos, addedTodo]);
+    } catch (error) {
+      setErrorMessage('Unable to add a todo');
+      throw error;
+    } finally {
+      setInputDisabled(false);
+      setTempTodo(null);
+    }
+  };
+
+  const filteredTodos = useMemo(() => todos.filter(({ completed }) => {
+    switch (condition) {
+      case ForComletedTodo.Active:
+        return !completed;
+      case ForComletedTodo.Completed:
+        return completed;
+      default:
+        return 1;
+    }
+  }), [condition, todos]);
+
+  const handleDeleteTodo = (todoId: number) => {
+    setInputDisabled(true);
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        throw new Error('Unable to delete a todo');
+      })
+      .finally(() => {
+        setInputDisabled(false);
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleRenameTodo = async (todo: Todo, newTodoTitle: string) => {
```

Comment:
this function and `handleToggleTodo` are very similar, maybe we can combine them

Reply:
this is still very difficult for me, I'll leave it